File: temporal-io/my-python/MY/0_simple/multiple_nodes.py
"""
Example 2: Multiple Nodes Workflow
This demonstrates a workflow with multiple activities executed sequentially.
Each activity represents a node in the workflow.
"""
import asyncio
from concurrent.futures import ThreadPoolExecutor
from datetime import timedelta

from temporalio import activity, workflow
from temporalio.client import Client
from temporalio.worker import Worker


# Activity 1: Validate input
@activity.defn
def validate_input(value: int) -> bool:
    activity.logger.info(f"Validating input: {value}")
    is_valid = value > 0
    activity.logger.info(f"Validation result: {is_valid}")
    return is_valid


# Activity 2: Transform data
@activity.defn
def transform_data(value: int) -> int:
    activity.logger.info(f"Transforming data: {value}")
    result = value * 3 + 10
    activity.logger.info(f"Transformed result: {result}")
    return result


# Activity 3: Save result
@activity.defn
def save_result(value: int) -> str:
    activity.logger.info(f"Saving result: {value}")
    return f"Successfully saved value: {value}"


# Workflow with multiple activity nodes
@workflow.defn
class MultipleNodesWorkflow:
    @workflow.run
    async def run(self, input_value: int) -> str:
        workflow.logger.info(f"Starting workflow with input: {input_value}")

        # Node 1: Validate input
        is_valid = await workflow.execute_activity(
            validate_input,
            input_value,
            start_to_close_timeout=timedelta(seconds=10),
        )

        if not is_valid:
            return "Input validation failed"

        # Node 2: Transform data
        transformed = await workflow.execute_activity(
            transform_data,
            input_value,
            start_to_close_timeout=timedelta(seconds=10),
        )

        # Node 3: Save result
        save_message = await workflow.execute_activity(
            save_result,
            transformed,
            start_to_close_timeout=timedelta(seconds=10),
        )

        workflow.logger.info(f"Workflow completed: {save_message}")
        return save_message


async def main():
    # Start client
    client = await Client.connect("localhost:7233")

    # Run a worker for the workflow
    async with Worker(
        client,
        task_queue="0-simple-multiple-nodes-task-queue",
        workflows=[MultipleNodesWorkflow],
        activities=[validate_input, transform_data, save_result],
        activity_executor=ThreadPoolExecutor(5),
    ):
        # Execute the workflow
        result = await client.execute_workflow(
            MultipleNodesWorkflow.run,
            15,
            id="0-simple-multiple-nodes-workflow",
            task_queue="0-simple-multiple-nodes-task-queue",
        )
        print(f"Workflow result: {result}")


if __name__ == "__main__":
    asyncio.run(main())

File: temporal-io/my-typescript/MY/0_simple/multiple_nodes.ts
/**
 * Example 2: Multiple Nodes Workflow
 * This demonstrates a workflow with multiple activities executed sequentially.
 *
 * This file contains:
 * - Activity definitions
 * - Workflow definitions
 * - Worker and workflow execution logic
 */

// ============================================================================
// ACTIVITIES
// ============================================================================
import { log as activityLog } from '@temporalio/activity';

/**
 * Activity 1: Validate input
 */
export async function validateInput(value: number): Promise<boolean> {
  activityLog.info('Validating input:', { value });
  const isValid = value > 0;
  activityLog.info('Validation result:', { isValid });
  return isValid;
}

/**
 * Activity 2: Transform data
 */
export async function transformData(value: number): Promise<number> {
  activityLog.info('Transforming data:', { value });
  const result = value * 3 + 10;
  activityLog.info('Transformed result:', { result });
  return result;
}

/**
 * Activity 3: Save result
 */
export async function saveResult(value: number): Promise<string> {
  activityLog.info('Saving result:', { value });
  return `Successfully saved value: ${value}`;
}

// ============================================================================
// WORKFLOWS
// ============================================================================
import { proxyActivities, log as workflowLog } from '@temporalio/workflow';

// Define activity interface
export interface Activities {
  validateInput(value: number): Promise<boolean>;
  transformData(value: number): Promise<number>;
  saveResult(value: number): Promise<string>;
}

// Proxy activities with timeout configuration
const {
  validateInput: validateInputActivity,
  transformData: transformDataActivity,
  saveResult: saveResultActivity,
} = proxyActivities<Activities>({
  startToCloseTimeout: '10 seconds',
});

/**
 * Workflow with multiple activity nodes executed sequentially
 */
export async function multipleNodesWorkflow(inputValue: number): Promise<string> {
  workflowLog.info('Starting workflow with input:', { inputValue });

  // Node 1: Validate input
  const isValid = await validateInputActivity(inputValue);

  if (!isValid) {
    return 'Input validation failed';
  }

  // Node 2: Transform data
  const transformed = await transformDataActivity(inputValue);

  // Node 3: Save result
  const saveMessage = await saveResultActivity(transformed);

  workflowLog.info('Workflow completed:', { saveMessage });
  return saveMessage;
}

// ============================================================================
// MAIN EXECUTION
// ============================================================================
import { Connection, Client } from '@temporalio/client';
import { NativeConnection, Worker } from '@temporalio/worker';
import { randomUUID } from 'crypto';

const TEMPORAL_HOST = 'localhost:7233';
const TASK_QUEUE = '0-simple-multiple-nodes-task-queue';

async function main() {
  try {
    // Connect to Temporal server
    const connection = await Connection.connect({ address: TEMPORAL_HOST });
    const client = new Client({ connection });
    console.log(`Connected to Temporal at ${TEMPORAL_HOST}`);

    // Create worker connection
    const workerConnection = await NativeConnection.connect({
      address: TEMPORAL_HOST,
    });

    // Create worker
    const worker = await Worker.create({
      connection: workerConnection,
      namespace: 'default',
      taskQueue: TASK_QUEUE,
      workflowsPath: new URL(import.meta.url).pathname,
      activities: {
        validateInput,
        transformData,
        saveResult,
      },
    });

    console.log(`Worker started on task queue: ${TASK_QUEUE}`);

    // Run worker in background
    const workerPromise = worker.run();

    // Give worker a moment to start
    await new Promise(resolve => setTimeout(resolve, 1000));

    // Execute workflow
    const workflowId = `multiple-nodes-workflow-${randomUUID()}`;
    console.log(`Starting workflow with ID: ${workflowId}`);

    const handle = await client.workflow.start(multipleNodesWorkflow, {
      args: [42],
      taskQueue: TASK_QUEUE,
      workflowId,
    });

    console.log('Workflow started, waiting for result...');
    const result = await handle.result();

    console.log(`Workflow completed successfully!`);
    console.log(`Workflow ID: ${workflowId}`);
    console.log(`Result: ${result}`);

    // Shutdown worker
    worker.shutdown();
    await workerPromise;
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

// Only run main if this file is executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}
